### PR TITLE
[DEVOPS-497] Kite

### PR DESCRIPTION
**Pull request by deepfire**
- a temporary fork of `buildkite-agent` -- we can't afford posting our secrets on github
- define a `kite` node that is able to connect to a buildkite account

### CODE REVIEW

File: modules/buildkite-agent.nix [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      { description = "Buildkite Agent";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        environment.HOME = "/var/lib/buildkite-agent";
+        preStart = ''
+            ${pkgs.coreutils}/bin/mkdir -m 0700 -p /var/lib/buildkite-agent/.ssh
+            ${copyOrEcho cfg.openssh.privateKey "/var/lib/buildkite-agent/.ssh/id_rsa"     600}
+            ${copyOrEcho cfg.openssh.publicKey  "/var/lib/buildkite-agent/.ssh/id_rsa.pub" 600}
+            cat > "/var/lib/buildkite-agent/buildkite-agent.cfg" <<EOF
+            token="${accessToken}"
```
Comment (domenkozar) [inline modules/buildkite-agent.nix]: Should be `$(cat "${accessToken}")`

Comment (cleverca22) [reply modules/buildkite-agent.nix]: i think it needs to be `$(cat "${toString accessTokenPath}")`

File: deployments/infrastructure-env-production.nix [hunk 2]
```diff
@@ -37,12 +37,28 @@ in {
     deployment.ec2.elasticIPv4 = resources.elasticIPs.cardanod-ip;
   };
 
+  kite = { config, pkgs, resources, ... }: {
```
Comment (domenkozar) [inline deployments/infrastructure-env-production.nix]: This will have to be part of each `hydra-slave` for proper integration, as the agents will run there.

File: deployments/infrastructure-target-aws.nix [hunk 3]
```diff
@@ -4,42 +4,34 @@ with (import ./../lib.nix);
...
-        resources.ec2SecurityGroups."allow-deployer-ssh-${region}-${org}"
-        resources.ec2SecurityGroups."allow-public-www-${region}-${org}"
-      ];
-    };
-  };
-
-  hydra-build-slave-1 = hydra;
-  hydra-build-slave-2 = hydra;
+    generic-infra-host =
+      { ebsInitialRootDiskSize
```
Comment (domenkozar) [inline deployments/infrastructure-target-aws.nix]: No function wrapping of NixOS modules anymore please :)

File: modules/buildkite-agent.nix [hunk 4]
```diff
@@ -0,0 +1,114 @@
+{ config, lib, pkgs, ... }:
```
Comment (domenkozar) [inline modules/buildkite-agent.nix]: I don't see what this file changes fundamentally that's needed for buildkite to work. All buildkite specific NixOS options already exist upstream.

File: nixpkgs-src.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
     "owner":  "NixOS",
     "repo":   "nixpkgs",
-    "rev": "6549173ae5dfece3c7f315185209cf4a54c2d257",
-    "sha256": "1lir7n1adyv8cm4g6c18qf4c2lmwcz085ra814m12hjgywp1i3my"
+    "rev": "b0e50cc7b97f96c7624bcabebf9ae5de51566225",
```
Comment (domenkozar) [inline nixpkgs-src.json]: This should be first upstreamed in separate PR (as discussed previously)

File: modules/buildkite-agent.nix (hunk 4 above)
Comment (deepfire) [reply modules/buildkite-agent.nix]: As discussed on the meeting, this allows provisioning secrets, without committing them to the Nix source.

File: nixpkgs-src.json (hunk 5 above)
Comment (deepfire) [reply nixpkgs-src.json]: Yes, this branch arrangement is temporary..

File: deployments/infrastructure-target-aws.nix (hunk 3 above)
Comment (deepfire) [reply deployments/infrastructure-target-aws.nix]: Oh, right : -)

By the way, can you suggest a simple way to achieve the same goal?

File: deployments/infrastructure-env-production.nix (hunk 2 above)
Comment (deepfire) [reply deployments/infrastructure-env-production.nix]: Understood..

File: modules/buildkite-agent.nix (hunk 4 above)
Comment (domenkozar) [reply modules/buildkite-agent.nix]: That's not required or important for the prototype issue, if I understand correctly.

File: deployments/infrastructure-target-aws.nix (hunk 3 above)
Comment (domenkozar) [reply deployments/infrastructure-target-aws.nix]: Add nixos configuraition as it was before, if you have a common setting, use `imports`

Comment (deepfire) [reply deployments/infrastructure-target-aws.nix]: Drat, that's what I was afraid to hear : -)

To be honest, personally this is hard to see as an improvement (splits related things away), so I'll just inline things back.

Comment (deepfire) [reply deployments/infrastructure-target-aws.nix]: Done!

File: modules/buildkite-agent.nix (hunk 4 above)
Comment (deepfire) [reply modules/buildkite-agent.nix]: We've addressed this in on Slack: this is needed to avoid leaking secrets.

File: deployments/infrastructure-env-production.nix (hunk 2 above)
Comment (deepfire) [reply deployments/infrastructure-env-production.nix]: This item is now tracked in https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-527

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+steps:
+  - label: 'iohk-ops'
+    command: 'scripts/ci.sh'
+    env:
+      NIX_REMOTE: daemon
+      PATH: bin:/run/current-system/sw/bin
```
Comment (domenkozar) [inline .buildkite/pipeline.yml]: Why is this needed? To find Nix?

Comment (deepfire) [reply .buildkite/pipeline.yml]: Yes, it wasn't finding `nix-shell`, since it was apparently clearing the environment,

Comment (domenkozar) [reply .buildkite/pipeline.yml]: It's not naive, it ensure isolation.

We should try getting rid of `PATH` and use `/run/current-system/sw/bin/nix-shell -p nixStable bash coreutils --run "scripts/ci.sh"` for the command

Comment (deepfire) [reply .buildkite/pipeline.yml]: Agreed to impurely reference `nix-shell`, but don't extend `PATH`.

File: deployments/infrastructure.nix [hunk 7]
```diff
@@ -3,12 +3,39 @@
...
+  mkHydraBuildSlave = { config, name, pkgs, ... }: {
     imports = [
       ./../modules/common.nix
       ./../modules/hydra-slave.nix
     ];
   };
+  mkBuildkiteAgent = { config, name, pkgs, ... }: {
+    imports = [
+      ./../modules/common.nix
+      ./../modules/hydra-slave.nix
```
[the remark this answers is not in the document]
Comment (deepfire) [reply deployments/infrastructure.nix]: Done.

File: deployments/infrastructure-target-aws.nix [hunk 8]
```diff
@@ -25,6 +25,9 @@ in rec {
   hydra-build-slave-1 = hydra;
   hydra-build-slave-2 = hydra;
 
+  buildkite-agent-1   = hydra;
```
[the remark this answers is not in the document]
Comment (deepfire) [reply deployments/infrastructure-target-aws.nix]: @domenkozar, I'm sorry from buildkite servers to where?

[the remark this answers is not in the document]
Comment (domenkozar) [reply deployments/infrastructure-target-aws.nix]: buildkite servers will ssh into our buildkite agent, so we have to allow SSH access

[the remark this answers is not in the document]
Comment (deepfire) [reply deployments/infrastructure-target-aws.nix]: I'm not quite sure why Buildkite servers will need to do that..

Do you think that our `iohk-ops` setup is missing some functionality that can be covered by some kind of extra Buildkite mechanism that relies on SSH-ing into our agents?

**Comment by deepfire** [comment]
Buildkite now passes -- ignoring the Travis failures, since there's no `.travis.yml`.
